File: src/oc_organizations_passwordpolicy/handlers.py
import logging
import sys
from uuid import uuid4
from typing import Any, Mapping, MutableMapping, Optional
from typing_inspect import is_union_type, get_args
from dataclasses import asdict, dataclass
from botocore.exceptions import ClientError

from cloudformation_cli_python_lib import (
    Action,
    HandlerErrorCode,
    OperationStatus,
    ProgressEvent,
    Resource,
    SessionProxy,
    exceptions,
)

from .models import ResourceHandlerRequest, ResourceModel

# Use this logger to forward log messages to CloudWatch Logs.
LOG = logging.getLogger(__name__)
TYPE_NAME = "OC::Organizations::PasswordPolicy"

resource = Resource(TYPE_NAME, ResourceModel)
test_entrypoint = resource.test_entrypoint

@dataclass
class PasswordPolicy(ResourceModel):
    def __setattr__(self, name: str, value: Any):
        anns = getattr(self, "__annotations__", {})
        type_attr = anns[name]
        if value is not None and type_attr:
            if is_union_type(type_attr):
                type_attr = get_args(type_attr)[0]
            if type_attr.__name__ == 'bool' and (value == 0 or value == 'false'):
                value = False
            elif type(value) is not type_attr:
                value = type_attr(value)
        super().__setattr__(name, value)

    def serialize(self) -> Mapping[str, Any]:
        return {k: v for k, v in asdict(self).items() if (v is not None and k != 'ResourceId' and k != 'ExpirePasswords')}

@resource.handler(Action.CREATE)
def create_handler(
    session: Optional[SessionProxy],
    request: ResourceHandlerRequest,
    callback_context: MutableMapping[str, Any],
) -> ProgressEvent:
    model = PasswordPolicy(**vars(request.desiredResourceState))
    progress: ProgressEvent = ProgressEvent(
        status=OperationStatus.IN_PROGRESS,
        resourceModel=model,
    )
    # try:
    #     if _retrieve_password_policy(session, request.desiredResourceState, request.logicalResourceIdentifier):
    #         raise Exception
    # except Exception:
    #     raise exceptions.AlreadyExists(TYPE_NAME, model.ResourceId or request.logicalResourceIdentifier)
    # except:
    #     pass
    if isinstance(session, SessionProxy):
        model = _upsert_password_policy(session, model, request.logicalResourceIdentifier)
        progress.resourceModel = model
        progress.status = OperationStatus.SUCCESS
    return progress

@resource.handler(Action.UPDATE)
def update_handler(
    session: Optional[SessionProxy],
    request: ResourceHandlerRequest,
    callback_context: MutableMapping[str, Any],
) -> ProgressEvent:
    model = PasswordPolicy(**vars(request.desiredResourceState))
    progress: ProgressEvent = ProgressEvent(
        status=OperationStatus.IN_PROGRESS,
        resourceModel=model,
    )
    _retrieve_password_policy(session, request.desiredResourceState, request.logicalResourceIdentifier)
    if isinstance(session, SessionProxy):
        model = _upsert_password_policy(session, model, model.ResourceId or request.logicalResourceIdentifier)
        progress.resourceModel = model  
        progress.status = OperationStatus.SUCCESS
    return progress

@resource.handler(Action.DELETE)
def delete_handler(
    session: Optional[SessionProxy],
    request: ResourceHandlerRequest,
    callback_context: MutableMapping[str, Any],
) -> ProgressEvent:
    model = request.desiredResourceState
    progress: ProgressEvent = ProgressEvent(
        status=OperationStatus.IN_PROGRESS,
        resourceModel=model,
    )
    progress = read_handler(session, request, callback_context)
    client = session.client('iam')
    client.delete_account_password_policy()
    LOG.info(f"{TYPE_NAME} [{progress.resourceModel.ResourceId}] [{request.logicalResourceIdentifier}] successfully deleted.")
    return progress

@resource.handler(Action.READ)
def read_handler(
    session: Optional[SessionProxy],
    request: ResourceHandlerRequest,
    callback_context: MutableMapping[str, Any],
) -> ProgressEvent:
    model = request.desiredResourceState
    progress: ProgressEvent = ProgressEvent(
        status=OperationStatus.IN_PROGRESS,
        resourceModel=model,
    )
    model = _retrieve_password_policy(session, model, request.logicalResourceIdentifier)
    progress.resourceModel = model
    progress.status = OperationStatus.SUCCESS
    return progress

@resource.handler(Action.LIST)
def list_handler(
    session: Optional[SessionProxy],
    request: ResourceHandlerRequest,
    callback_context: MutableMapping[str, Any],
) -> ProgressEvent:
    progress: ProgressEvent = ProgressEvent(
        status=OperationStatus.IN_PROGRESS,
        resourceModels=[],
    )
    model = _retrieve_password_policy(session, request.desiredResourceState, request.logicalResourceIdentifier)
    progress.resourceModels = [model]
    progress.status = OperationStatus.SUCCESS
    return progress

def _retrieve_password_policy(
    session: SessionProxy,
    model: Mapping[str, Any],
    logical_resource_identifier: Optional[str],
) -> PasswordPolicy:
    try:
        client = session.client('iam')
        response = client.get_account_password_policy()
        merged = {**vars(model), **response['PasswordPolicy']}
        model = PasswordPolicy(**merged)
        if model.ResourceId is None:
            model.ResourceId = logical_resource_identifier
        LOG.info(f"{TYPE_NAME} [{model.ResourceId}] [{logical_resource_identifier}] successfully retrieved.")
    except ClientError as e:
        if e.response.get('Error', {}).get('Code') == 'NoSuchEntity':
            raise exceptions.NotFound(TYPE_NAME, model.ResourceId or logical_resource_identifier)
        else: # raise the original exception
            raise
    return model

def _upsert_password_policy(
    session: SessionProxy,
    model: PasswordPolicy,
    logical_resource_identifier: Optional[str],
) -> PasswordPolicy:
    try:
        client = session.client('iam')
        if model.ResourceId is None:
            model.ResourceId = str(uuid4())
        client.update_account_password_policy(**model.serialize())
        LOG.info(f"{TYPE_NAME} [{model.ResourceId}] [{logical_resource_identifier}] successfully upserted.")
    except TypeError as e:
        raise exceptions.InternalFailure(f"was not expecting type {e}")
    return model

File: src/handlers.ts
import 'reflect-metadata';
import {
    jsonObject,
    jsonMember,
    toJson,
    TypedJSON,
} from 'typedjson';
import { config, IAM } from 'aws-sdk';
import { v4 as uuidv4 } from 'uuid';
import {
    Action,
    BaseResource,
    exceptions,
    handlerEvent,
    OperationStatus,
    Optional,
    ProgressEvent,
    ResourceHandlerRequest,
    SessionProxy,
} from 'cfn-rpdk';

import { ResourceModel } from './models';

// Use this logger to forward log messages to CloudWatch Logs.
const LOGGER = console;

/*
// The following can be used for debugging AWS SDK
config.logger = LOGGER;
Error.stackTraceLimit = Infinity;
*/

const parseBoolean = (value: string | boolean): boolean => {
    if (value !== undefined && value !== null) {
        return !!JSON.parse(String(value).toLowerCase());
    }
};

const parseInteger = (value: string | number): number => {
    if (value !== undefined && value !== null) {
        if (typeof value === 'string') {
            return Number.parseInt(value, 10);
        } else if (typeof value === 'number') {
            return Math.trunc(value);
        }
    }
};

@jsonObject
@toJson
class PasswordPolicy extends ResourceModel {

    @jsonMember({ constructor: String })
    ResourceId: Optional<string>;
    @jsonMember({ deserializer: parseInteger, constructor: Number })
    MinimumPasswordLength: Optional<number>;
    @jsonMember({ deserializer: parseBoolean, constructor: Boolean })
    RequireSymbols: Optional<boolean>;
    @jsonMember({ deserializer: parseBoolean, constructor: Boolean })
    RequireNumbers: Optional<boolean>;
    @jsonMember({ deserializer: parseBoolean, constructor: Boolean })
    RequireUppercaseCharacters: Optional<boolean>;
    @jsonMember({ deserializer: parseBoolean, constructor: Boolean })
    RequireLowercaseCharacters: Optional<boolean>;
    @jsonMember({ deserializer: parseBoolean, constructor: Boolean })
    AllowUsersToChangePassword: Optional<boolean>;
    ExpirePasswords: Optional<boolean>;
    @jsonMember({ deserializer: parseInteger, constructor: Number })
    MaxPasswordAge: Optional<number>;
    @jsonMember({ deserializer: parseInteger, constructor: Number })
    PasswordReusePrevention: Optional<number>;
    @jsonMember({ deserializer: parseBoolean, constructor: Boolean })
    HardExpiry: Optional<boolean>;

    public static fromObject(input: object): PasswordPolicy {
        const serializer = new TypedJSON(PasswordPolicy);
        return serializer.parse(input);
    }

    public toObject(): any {
        const serializer = new TypedJSON(PasswordPolicy);
        return JSON.parse(JSON.stringify(serializer.toPlainJson(this)));
    }
}

/**
 * Retrieves the password policy for the AWS account.
 * 
 * @param session Current AWS session passed through from caller
 * @param resourceId The resource unique identifier
 * @param logicalResourceId The logical name of the resource as specified
 * in the template
 */
const retrievePasswordPolicy = async (
    session: SessionProxy,
    resourceId: Optional<string>,
    logicalResourceId: Optional<string>
): Promise<PasswordPolicy> => {
    let model: PasswordPolicy = null;
    try {
        if (session instanceof SessionProxy) {
            const client = session.client('IAM') as IAM;
            const response = await client.getAccountPasswordPolicy().promise();
            // LOGGER.debug('getAccountPasswordPolicy response', response);
            model = PasswordPolicy.fromObject(response.PasswordPolicy);
            model.ResourceId = resourceId || logicalResourceId;
            LOGGER.info(
                PasswordPolicy.TYPE_NAME,
                `[${model.ResourceId}] [${logicalResourceId}]`,
                'successfully retrieved.'
            );
        }
    } catch(err) {
        LOGGER.log(err);
        if (err && err.code === 'NoSuchEntity') {
            throw new exceptions.NotFound(
                PasswordPolicy.TYPE_NAME,
                resourceId || logicalResourceId
            );
        } else { // Raise the original exception
            throw err;
        }
    }
    return Promise.resolve(model);
};

/**
 * Updates or insert the password policy settings for the AWS account. This operation does not
 * support partial updates. No parameters are required, but if you do not specify a
 * parameter, that parameter's value reverts to its default value.
 * 
 * @param session Current AWS session passed through from caller
 * @param model The updated options for the password policy
 * @param logicalResourceId The logical name of the resource as specified
 * in the template
 */
const upsertPasswordPolicy = async (
    session: SessionProxy,
    model: PasswordPolicy,
    logicalResourceId: Optional<string>
): Promise<PasswordPolicy> => {
    try {
        if (session instanceof SessionProxy) {
            const client = session.client('IAM') as IAM;
            const params = model.toObject();
            delete params['ResourceId'];
            // LOGGER.debug('updateAccountPasswordPolicy params', params);
            const response = await client.updateAccountPasswordPolicy(params).promise();
            // LOGGER.debug('updateAccountPasswordPolicy response', response);
            if (!model.ResourceId) {
                model.ResourceId = uuidv4();
            }
            LOGGER.info(
                PasswordPolicy.TYPE_NAME,
                `[${model.ResourceId}] [${logicalResourceId}]`,
                'successfully upserted.'
            );
        }
    } catch(err) {
        LOGGER.log(err);
        throw new exceptions.InternalFailure(err.message);
    }
    return Promise.resolve(model);
}

class Resource extends BaseResource<ResourceModel> {

    /**
     * CloudFormation invokes this handler when the resource is initially created
     * during stack create operations.
     * 
     * @param session Current AWS session passed through from caller
     * @param request The request object for the provisioning request passed to the implementor
     * @param callbackContext Custom context object to enable handlers to process re-invocation
     */
    @handlerEvent(Action.Create)
    public async create(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>
    ): Promise<ProgressEvent> {
        request.desiredResourceState = PasswordPolicy.fromObject(request.desiredResourceState);
        request.previousResourceState = PasswordPolicy.fromObject(request.previousResourceState);
        // LOGGER.debug('CREATE request', request);
        let model: PasswordPolicy = request.desiredResourceState;
        const progress: ProgressEvent<PasswordPolicy> = ProgressEvent.builder()
            .status(OperationStatus.InProgress)
            .resourceModel(model)
            .build() as ProgressEvent<PasswordPolicy>;
        model = await upsertPasswordPolicy(session, model, request.logicalResourceIdentifier);
        progress.resourceModel = model;
        progress.status = OperationStatus.Success;
        LOGGER.info('CREATE progress', progress.toObject());
        return progress;
    }

    /**
     * CloudFormation invokes this handler when the resource is updated
     * as part of a stack update operation.
     * 
     * @param session Current AWS session passed through from caller
     * @param request The request object for the provisioning request passed to the implementor
     * @param callbackContext Custom context object to enable handlers to process re-invocation
     */
    @handlerEvent(Action.Update)
    public async update(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>
    ): Promise<ProgressEvent> {
        request.desiredResourceState = PasswordPolicy.fromObject(request.desiredResourceState);
        request.previousResourceState = PasswordPolicy.fromObject(request.previousResourceState);
        // LOGGER.debug('UPDATE request', request);
        let model: PasswordPolicy = request.desiredResourceState;
        const progress: ProgressEvent<PasswordPolicy> = ProgressEvent.builder()
            .status(OperationStatus.InProgress)
            .resourceModel(model)
            .build() as ProgressEvent<PasswordPolicy>;
        await retrievePasswordPolicy(
            session,
            model.ResourceId,
            request.logicalResourceIdentifier
        );
        model = await upsertPasswordPolicy(session, model, request.logicalResourceIdentifier);
        // LOGGER.debug('UPDATE model', model);
        progress.resourceModel = model;
        progress.status = OperationStatus.Success;
        // progress.callbackContext = new Map<string, any>();
        // callbackContext.set('RequestId', 'e9899120-8e56-11ea-bfb5-02722d612c06');
        LOGGER.info('UPDATE progress', progress.toObject());
        return progress;
    }

    /**
     * CloudFormation invokes this handler when the resource is deleted, either when
     * the resource is deleted from the stack as part of a stack update operation,
     * or the stack itself is deleted.
     * 
     * @param session Current AWS session passed through from caller
     * @param request The request object for the provisioning request passed to the implementor
     * @param callbackContext Custom context object to enable handlers to process re-invocation
     */
    @handlerEvent(Action.Delete)
    public async delete(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>
    ): Promise<ProgressEvent> {
        request.desiredResourceState = PasswordPolicy.fromObject(request.desiredResourceState);
        request.previousResourceState = PasswordPolicy.fromObject(request.previousResourceState);
        // LOGGER.debug('DELETE request', request);
        let model: PasswordPolicy = request.desiredResourceState;
        const progress: ProgressEvent<PasswordPolicy> = ProgressEvent.builder()
            .status(OperationStatus.InProgress)
            .resourceModel(model)
            .build() as ProgressEvent<PasswordPolicy>;
        model = await retrievePasswordPolicy(
            session,
            model.ResourceId,
            request.logicalResourceIdentifier
        );
        try {
            if (session instanceof SessionProxy) {
                const client = session.client('IAM') as IAM;
                const response = await client.deleteAccountPasswordPolicy().promise();
                // LOGGER.debug('deleteAccountPasswordPolicy response', response);
                LOGGER.info(
                    this.typeName,
                    `[${model.ResourceId}] [${request.logicalResourceIdentifier}]`,
                    'successfully deleted.'
                );
            }
        } catch(err) {
            LOGGER.log(err);
            throw new exceptions.InternalFailure(err.message);
        }
        progress.status = OperationStatus.Success;
        LOGGER.info('DELETE progress', progress.toObject());
        return progress;
    }

    /**
     * CloudFormation invokes this handler as part of a stack update operation when
     * detailed information about the resource's current state is required.
     * 
     * @param session Current AWS session passed through from caller
     * @param request The request object for the provisioning request passed to the implementor
     * @param callbackContext Custom context object to enable handlers to process re-invocation
     */
    @handlerEvent(Action.Read)
    public async read(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>
    ): Promise<ProgressEvent> {
        request.desiredResourceState = PasswordPolicy.fromObject(request.desiredResourceState);
        request.previousResourceState = PasswordPolicy.fromObject(request.previousResourceState);
        // LOGGER.debug('READ request', request);
        const model: PasswordPolicy = await retrievePasswordPolicy(
            session,
            request.desiredResourceState.ResourceId,
            request.logicalResourceIdentifier
        );
        const progress: ProgressEvent<PasswordPolicy> = ProgressEvent.builder()
            .status(OperationStatus.Success)
            .resourceModel(model)
            .build() as ProgressEvent<PasswordPolicy>;
        LOGGER.info('READ progress', progress.toObject());
        return progress;
    }

    /**
     * CloudFormation invokes this handler when summary information about multiple
     * resources of this resource provider is required.
     * 
     * @param session Current AWS session passed through from caller
     * @param request The request object for the provisioning request passed to the implementor
     * @param callbackContext Custom context object to enable handlers to process re-invocation
     */
    @handlerEvent(Action.List)
    public async list(
        session: Optional<SessionProxy>,
        request: ResourceHandlerRequest<ResourceModel>,
        callbackContext: Map<string, any>
    ): Promise<ProgressEvent> {
        request.desiredResourceState = PasswordPolicy.fromObject(request.desiredResourceState);
        request.previousResourceState = PasswordPolicy.fromObject(request.previousResourceState);
        // LOGGER.debug('LIST request', request);
        const models: Array<PasswordPolicy> = [];
        try {
            const model: PasswordPolicy = await retrievePasswordPolicy(
                session,
                request.desiredResourceState.ResourceId,
                request.logicalResourceIdentifier
            );
            models.push(model);
        } catch(err) {
            if (!(err instanceof exceptions.NotFound)) {
                throw err;
            }
        }
        const progress: ProgressEvent<PasswordPolicy> = ProgressEvent.builder()
            .status(OperationStatus.Success)
            .resourceModels(models)
            .build() as ProgressEvent<PasswordPolicy>;
        LOGGER.info('LIST progress', progress.toObject());
        return progress;
    }
}

const resource = new Resource(PasswordPolicy.TYPE_NAME, PasswordPolicy);

export const entrypoint = (...args: [any, any]) => {
    // LOGGER.debug('entrypoint input', ...args);
    return resource.entrypoint(...args);
}

export const testEntrypoint = (...args: [any, any]) => {
    // LOGGER.debug('testEntrypoint input', ...args);
    return resource.testEntrypoint(...args);
}
